Add dev policy document for testing standards

james-stocks opened the pull request:
Following [Chef's OSS design process](https://github.com/chef/chef-oss-practices/blob/master/contributors/guide/design-proposals.md) this PR fixes #2473 and finalises the design.

Any feedback on the contents of this design must be left as a comment on #2473 so we keep the discussion in a single place.
Please only add comments to this PR for critique of files structure/formatting, spelling/punctuation/grammar, etc - do not add comments for the semantics of the text.

There are some changes between the text in #2473 and what's being submitted here:
1. I deleted the "`tests` folder structure" section as most of it was no longer applicable due to changes agreed during the design windows.
2. Changed the word "should" to "will" in a number of places to be more direct about what will happen.
3. Some small corrections and whitespace fixes.

gavindidrichsen commented on docs/dev/policy_documents/standards-for-tests.md:
```diff
@@ -0,0 +1,144 @@
+#### Test implementation
+
+* The standard testing programs used for core-plans are [BATS](https://github.com/sstephenson/bats) (for Linux) and [Pester](https://github.com/pester/Pester) (for Windows).
+* Tests should not execute inside the studio instance than the package was built in; to avoid cases where the software works only due to build dependencies being present.
```
Should read "...instance **that** the package..." instead of "...than..."

gavindidrichsen replied:
Should read "...instance that the package..." instead of "...than..."

james-stocks commented:
@predominant - good call.

Some current examples (not yet merged) will be:

[sed] https://github.com/habitat-sh/core-plans/pull/2401
[dotnet-471-dev-pack] https://github.com/habitat-sh/core-plans/pull/2443

I will copy those examples into this document

james-stocks commented:
@predominant - 1x Windows and 1x Linux example added to the document as an Appendix

james-stocks commented:
Change needed:
 * the examples given break the design proposed because they run the tests in the same studio the package was built it.

Possible improvements:
 * `PKG_IDENT` is a bad variable name because `pkg_ident`is already a variable used in habitat.
 * The Windows example does not show testing a program, it only asserts on folder existence. An example of testing a program would be nice.

james-stocks commented:
Hi @predominant -
I have updated the examples to change them to 7zip (which has both a Linux and Windows plan and the tests can be directly comparable).
I changed the name of the var we use for the package ident to `$TEST_PACKAGE_IDENT` to make it more self-descriptive and to avoid it clashing with the existing `pkg_ident`.
I also changed the example test steps to demonstrate _not_ running the tests in the same studio instance as the build.

predominant commented:
Examples of the one-liners can be seen in #2553
